cpp/arcticdb/storage/lmdb/lmdb_storage.cpp

```diff
@@ -115,10 +115,13 @@ void LmdbStorage::do_read(Composite<VariantKey>&& ks, const ReadVisitor& visitor
             MDB_val mdb_val;
             ARCTICDB_SUBSAMPLE(LmdbStorageGet, 0)
 
-            if (::lmdb::dbi_get(txn, dbi.handle(), &mdb_key, &mdb_val)) {
+            if (::lmdb::dbi_get(*txn, dbi.handle(), &mdb_key, &mdb_val)) {
                 ARCTICDB_SUBSAMPLE(LmdbStorageVisitSegment, 0)
-                visitor(k, Segment::from_bytes(reinterpret_cast<std::uint8_t *>(mdb_val.mv_data),
-                                               mdb_val.mv_size));
+                auto segment = Segment::from_bytes(reinterpret_cast<std::uint8_t *>(mdb_val.mv_data),
+                                                   mdb_val.mv_size);
+                std::any keepalive;
+                segment.set_keepalive(std::any(std::move(txn)));
```

Comment:
Do we have to now be careful that `segment` is not used by others threads, and that this thread does not open another transaction? See http://www.lmdb.tech/doc/group__mdb.html#gad7ea55da06b77513609efebd44b26920 `mdb_txn_begin` docs:

```
A transaction and its cursors must only be used by a single thread, and a thread may only have a single transaction at a time. If [MDB_NOTLS](http://www.lmdb.tech/doc/group__mdb__env.html#ga5dca84a576d14b4bfe2deddc2dc622d3) is in use, this does not apply to read-only transactions. 
    Cursors may not span transactions.
```

Reply:
We're fine as long as they're read-only transactions. Also we set up the environment with MDB_NOTLS, so the reader locktable isn't tied to a particular thread.